File: backjoon_level_python/1079.py
def solve(N, guilts, matrix):

    def dfs(num_people_left, night_count):
        global maximum_night_count
        maximum_night_count = max(maximum_night_count, night_count)

        # 낮
        day_flag = False
        if num_people_left % 2 != 0:
            day_flag = True

            maximum_guilt_idx = -1
            temp_max = -1
            for idx, guilt in enumerate(guilts):
                if visited[idx] is False and guilt > temp_max:
                    temp_max = guilt
                    maximum_guilt_idx = idx

            if maximum_guilt_idx == Eunjin:
                return

            # kill maximum_guilt_idx
            visited[maximum_guilt_idx] = True
            num_people_left -= 1
        # 밤
        for i in range(N):
            if i != Eunjin and visited[i] is False:
                visited[i] = True # Kill Person i
                num_people_left -= 1

                # 다른 참가자 j의 유죄 지수는 R[i][j]만큼 변한다.
                for j in range(len(guilts)):
                    guilts[j] += matrix[i][j]

                dfs(num_people_left, night_count + 1)

                for j in range(len(guilts)):
                    guilts[j] -= matrix[i][j]
                visited[i] = False
                num_people_left += 1

        # 재귀 끝날때 낮에 죽은사람 다시 살려야함
        if day_flag:
            visited[maximum_guilt_idx] = False
            num_people_left += 1

    dfs(N, 0)
    return maximum_night_count

if __name__ == '__main__':
    N = int(input())
    visited = [False] * N

    guilts = list(map(int, input().split()))
    matrix = [list(map(int, input().split())) for _ in range(N)]

    Eunjin = int(input())

    maximum_night_count = 0

    print(solve(N, guilts, matrix))
    # solve(4, [500]*4, [[1,4,3,-2],[-2,1,4,3],[3,-2,1,4],[4,3,-2,1]])

File: backjoon_level_c/backjoon_level_c/1079.cpp
#include <iostream>
using namespace std;

int N, eunjin;
int ans = -1;

int guilt[16];
int R[16][16];
bool visit[16];

// 유죄점수가 가장 높은 idx구하기, 같으면 낮은 idx
int calculate()
{
    int idx = -1;
    int idx_point = -10000;
 
    for (int i = 0; i < N; i++)
    {
        if (guilt[i] > idx_point && !visit[i])
        {
            idx = i;
            idx_point = guilt[i];
        }
    }
    return idx;
}

void dfs(int numPeopleLeft, int nightCount){
    if(nightCount > ans){
        ans = nightCount;
    }

    int idx = -1;
    if(numPeopleLeft % 2 != 0){
        idx = calculate();
        if(idx == eunjin){
            return;
        }
        visit[idx] = true;
        --numPeopleLeft;
    }
    // 밤
    for(int i=0;i<N;i++){
        if(visit[i] == false){
            visit[i] = true;
            --numPeopleLeft;

            for(int j=0;j<N;j++){
                guilt[j] += R[i][j];
            }
            dfs(numPeopleLeft, nightCount + 1);
            for(int j=0;j<N;j++){
                guilt[j] -= R[i][j];
            }
            
            ++numPeopleLeft;
            visit[i] = false;
        }
    }
    // 재귀 끝날때 낮에 죽은사람 다시 살려야함
    if(numPeopleLeft % 2 != 0){
        visit[idx] = false;
        ++numPeopleLeft;
    }
}
int main(){
    cin >> N;

    for (int i = 0; i < N; i++)
        cin >> guilt[i];
    
    for (int i = 0; i < N; i++)
        for (int j = 0; j < N; j++)
            cin >> R[i][j];
    
    cin >> eunjin;
    visit[eunjin] = true;
    dfs(N, 0);

    cout << ans;

    return 0;
}